### ISSUE TITLE
Is there a way to put Elmo as a Keras layer and integrate it into a Keras model?

### ISSUE THREAD

**Issue by jacklxc** [allenai/bilm-tf#73]
If there is a way to put Elmo as a Keras layer and integrate it into a Keras model, could you please provide an example just like `usage_character.py`? Thank you very much.

**Comment by matt-peters** [comment]
This would be a great addition, but we don't have an example like this.

**Comment by msiahbani** [comment]
@jacklxc any updates? I would appreciate if you let me know whether you could get it done. I am gonna use it in a different framework and any example is appreciated! thanks.

**Comment by kermitt2** [comment]
Hello! 
I integrated ELMo in my Keras framework, you can have a look here: 
https://github.com/kermitt2/delft

In Keras, I think the simplest way to integrate "dynamic" embeddings is to use  `fit_generator` from `keras.utils.Sequence` rather than creating a layer. It is really designed for this and offers a lot of flexibility - you can create on the fly ad hoc embeddings easily. This is useful here because we might want to concatenate ELMo with Glove embeddings for instance because it gives an interesting performance boost. 

My Keras data generator is here: https://github.com/kermitt2/delft/blob/master/sequenceLabelling/data_generator.py

Another comment, in my implementation, I cache the ELMo embeddings for a complete sequence so that the same ELMo embeddings are not re-calculated for each epoch during training, but simply on the first epoch (and at test time of course). I use LMDB memory page for this, so that the embeddings are not loaded in memory (simlarly as the way LMDB is used by Torch or Caffe).

**Comment by jacklxc** [comment]
@kermitt2 Thank you for your great work! I checked your code and found that essentially your are using Tensorflow code provided by Elmo to compute embeddings. Then you put the embeddings as inputs to the Keras model. In that case, Elmo weights are not trainable during the training of Keras model. 

I wonder if there is anyway to make Tensorflow implementation of Elmo as a trainable embedding layer. [https://towardsdatascience.com/elmo-embeddings-in-keras-with-tensorflow-hub-7eb6f0145440](url) gives a good example, but the embedding weights are from Tensorflow Hub, which is pre-trained. In my case, I want to use my own trained embedding weights.

**Comment by nreimers** [comment]
I added some examples here:
https://github.com/UKPLab/elmo-bilstm-cnn-crf
https://github.com/UKPLab/elmo-bilstm-cnn-crf/blob/master/Keras_ELMo_Tutorial.ipynb

There I used ELMo to train a Keras network (for sentence classification and sequence tagging).

In my first experiments I had the ELMo layer as a trainable embedding layer. However, this produces computationally quite a large overhead, because the ELMo embeddings must be computed each time. Computing the embeddings once and fixing them gave a big performance boost (at training time) without reducing the performance.

**Comment by kermitt2** [comment]
@jacklxc We normally don't want to update the word embeddings when using pre-trained embeddings to avoid overfitting and to reduce the size of the model. What would be the particular motivation for doing this? In the case of ELMo, as the embeddings are provided by a separate model (it's not just a look-up table), I think it would mean to integrate it entirely with the keras model (weights included), and writing our own Keras layer that would update the ELMo weights based on the input tensor - complicated I think. 

The solution proposed in https://towardsdatascience.com/elmo-embeddings-in-keras-with-tensorflow-hub-7eb6f0145440 does not make the embedding layer trainable with the Keras model training imho, it simply mirrors the tensor output of the tensor hub module. Lambda is a stateless operation in Keras. If you look at the corresponding https://github.com/strongio/keras-elmo/blob/master/Elmo%20Keras.ipynb
you can see the the Embedding layer (lambda_1) has 0 trainable parameters. 

@nreimers It's not because the ELMo embeddings are recomputed at each epoch that the embeddings are trainable. Your solution is equivalent to mine, ELMo embeddings are obtained "externally from Keras"   (with AllenNLP ElmoEmbedder in your case, with BiLM-tf library in my case), but not trainable with the Keras model (which is good I think when integrating ELMo in Keras!). 

I might have misunderstood, but in your implementation, I have the impression that the whole embeddings are computed in advance for all the sentences of the dataset (method `sentenceLookup` in `ELMoWordEmbeddings.py`). This will blow the memory as soon the training data is a bit large. For instance with the Kaggle toxic comment classification challenge, where we have 150,000 comments with an average of 200 words each, 1024 dim, 4 bytes per value, we would reach more than 120 GB of embeddings to put in memory (ELMo embeddings alone). The way to solve this problem with Keras is to use DataGenerator as I pointed above. But if I misread your code, please apologize me! 

For caching ELMo embeddings during training, I suggest LMDB rather than file system, which really gives an even better boost and generalizes nicely to usual embeddings.

Your mini-batch method with similar sentence length is really nice!

**Comment by nreimers** [comment]
@kermitt2 You are right, in my implementation the ELMo embeddings are pre-computed for all sentences and then stored on disk  (and hold in RAM while training).

This is of course not suitable for large datasets, as it will require too much RAM if your dataset is too large.

The released code is experimental code I use in my sequence tagging experiments. In sequence tagging experiments, datasets are usually smaller compared to text classification experiments. For example, the ELMo embeddings for the CoNLL 2003 NER dataset requires about 1 GB of RAM, which is feasible even for desktop computers to store in RAM.

I didn't use a generator as I must train a lot of epochs (and with many different configurations), so computing the embeddings once and storing them is the most compute time efficient. It's the classical CPU-RAM trade-off: Less CPU by pre-computing the ELMo embeddings once but higher storage/RAM requirements. As the sequence tagging datasets are often small (the largest dataset I have requires about 2 GB of RAM to store the ELMo embeddings for all sentences and that dataset is considered as quite large), storing all ELMo embeddings in RAM is not an issue (for me).
 
But of course, once your datasets get larger, you need to use a DataGenerator at some point.

**Comment by jbrry** [comment]
@nreimers 

> ... in my implementation the ELMo embeddings are pre-computed for all sentences and then stored on disk (and hold in RAM while training).

I am working on a project which is implemented in Python 2, as such we are planning to compute ELMo representations for the datasets (train/dev/test) as an external process. So far I have converted the dataset into a text file which contains each sentence line by line like in the format listed on the [ELMo instructions](https://github.com/allenai/allennlp/blob/master/tutorials/how_to/elmo.md), e.g.:

`echo "The cryptocurrency space is now figuring out to have the highest search on Google globally ." > sentences.txt` 

I then issue the following command from within Python on my data, e.g. 'train_sentences.txt' file: 

`allennlp elmo train_sentences.txt elmo_layers.hdf5 --all`

This works fine and passes the checks, e.g. the vector for a sentence has the format [3, <# of tokens in sent>, 1024].

So basically my question is, is this what you do in your repository when you say you pre-compute the representations for all sentences? If that is the case, do you know the best way to load in the vectors (from disk) into your main repository (e.g. a Python 2 repo) and then add them to the relative sentences? I think in your repository you keep track of a sentence index. I've taken a look at your repository but I am not sure if our approaches are similar. 

Thanks

**Comment by nreimers** [comment]
Hi @Jbar-ry 
I use a slightly different approach. Instead of running the allennlp binary, is use AllenNLP as a framework to create the ELMo embeddings and then storing them.

Have a look in this file:
https://github.com/UKPLab/elmo-bilstm-cnn-crf/blob/master/Create_ELMo_Cache.py

The code reads a text file and create a dictionary lookup, that maps sentence -> ELMo embeddings. This avoids that an identical sentence is computed twice.

I store this dict using pickle.

However, this is only advisable when your text file is small enough. At least in the past, pickle hat some issues storing files larger than 4 GB (not sure if this is fixed already).

**Comment by jbrry** [comment]
Hi @nreimers, 

Thanks very much for the response. I will try the method you've used in your repository, thanks!

**Comment by jacobzweig** [comment]
Hi - just a small note that I've updated the example in https://github.com/strongio/keras-elmo/blob/master/Elmo%20Keras.ipynb to allow trainable weights. Hope that helps :-)

**Comment by IS5882** [comment]
@jacobzweig Hi Jacob, thanks for your trainable Elmo 👍 , but any ideas why when I change the signature to output a 3-D tensor from 'default' to 'Elmo' the output is still 2-D? I did exactly as shown here [(https://tfhub.dev/google/elmo/2)] (https://tfhub.dev/google/elmo/2) 
Thanks

**Comment by gupta-rajat** [comment]
If I wanted to use "word_emb" instead of "default" for the output. I tried to use Elmo embedding layer in the model but  I got some error. Please if somebody can look into this.
https://stackoverflow.com/questions/55927850/elmo-embedding-layer-by-using-keras-layer

**Comment by ghost** [comment]
@jacobzweig 
hi . first of fall thank you for [the code]( https://github.com/strongio/keras-elmo/blob/master/Elmo%20Keras.ipynb) 
but the question is ,is there anyway to not use TensorFlow hub in this code and load my own trained model (hd5 weight file and option file)?(i trained my model with below code:
https://github.com/allenai/bilm-tf/)
and how?
I cant find any code to support this kind of loading
thank you in advance!

**Comment by harshitjain94** [comment]
Looking for same query

@jacobzweig
hi . first of fall thank you for the code
but the question is ,is there anyway to not use TensorFlow hub in this code and load my own trained model (hd5 weight file and option file)?(i trained my model with below code:
https://github.com/allenai/bilm-tf/)
and how?
I cant find any code to support this kind of loading
thank you in advance!

**Comment by nreimers** [comment]
@harshitjain94 
Maybe this trick can help you:
https://github.com/UKPLab/elmo-bilstm-cnn-crf/blob/master/Keras_ELMo_Tutorial.ipynb

There, I use AllenNLP to generate ELMo embeddings for each word. These embeddings are than passed to Keras. 

The only "downside" is, that the ELMo network / embeddings cannot be updated during the training. But in my experiments (https://arxiv.org/abs/1904.02954), this updating of the ELMo network is not really needed (not improving the performance significantly for the tasks I tested).

You can load any ELMo model into Keras with this method that you could use with AllenNLP.

Best regards
Nils Reimers
